qiita_db/analysis.py

```diff
@@ -802,59 +805,69 @@ def build_files(self, rarefaction_depth=None):
-            # each data type
-            new_tables = {dt: None for dt in self.data_types}
             base_fp = qdb.util.get_work_base_dir()
-            for a_id, samps in viewitems(samples):
-                # one biom table attached to each artifact object
-                artifact = qdb.artifact.Artifact(a_id)
-                table_fp = None
+
+            # this assumes that there is only one refererence/pipeline for each
+            # data_type issue #164
+            new_tables = {dt: None for dt in self.data_types}
+            for aid, samps in viewitems(samples):
+                artifact = qdb.artifact.Artifact(aid)
+                # this is not checking the reference used for picking
+                # issue #164
+                biom_table_fp = None
                 for _, fp, fp_type in artifact.filepaths:
                     if fp_type == 'biom':
-                        table_fp = fp
+                        biom_table_fp = fp
                         break
-                if not table_fp:
+                if not biom_table_fp:
                     raise RuntimeError(
                         "Artifact %s do not have a biom table associated"
-                        % a_id)
-                table = load_table(table_fp)
-                # HACKY WORKAROUND FOR DEMO. Issue # 246
-                # make sure samples not in biom table are not filtered for
-                table_samps = set(table.ids())
-                filter_samps = table_samps.intersection(samps)
-                # add the metadata column for study the samples come from
-                study_meta = {'Study': artifact.study.title,
-                              'Processed_id': artifact.id}
-                samples_meta = {sid: study_meta for sid in filter_samps}
-                # filter for just the wanted samples and merge into new table
-                # this if/else setup avoids needing a blank table to
-                # start merges
-                table.filter(filter_samps, axis='sample', inplace=True)
-                table.add_metadata(samples_meta, axis='sample')
+                        % aid)
+                biom_table = load_table(biom_table_fp)
+                # filtering samples to keep those selected by the user
+                biom_table_samples = set(biom_table.ids())
+                selected_samples = biom_table_samples.intersection(samps)
+                biom_table.filter(selected_samples, axis='sample',
+                                  inplace=True)
+
+                if not duplicated_samples:
+                    ids_map = {_id: "%d.%s" % (aid, _id)
+                               for _id in biom_table.ids()}
+                    biom_table.update_ids(ids_map, 'sample', True, True)
+
+                # add the metadata column for study the samples come from,
+                # this is useful in case the user download the bioms
+                study_md = {'Study': artifact.study.title, 'Processed_id': aid}
```

Comment:
Yup, note that this was a left over from when we moved to artifacts and didn't change to not move things around. Changing now.